enhancements/cassandra-configuration.md

```diff
@@ -0,0 +1,265 @@
+  serverVersion: 3.11.7
+  managementApiAuth:
+    insecure: {}
+  size: 9
+  racks:
+  - name: rack-1
+    zone: us-east1-b
+  - name: rack-2
+    zone: us-east1-c
+  - name: rack-3
+    zone: us-east1-d
+  ...
+```
+
+**Note:** Zones will soon be deprecated in favor of free form labels.
+
+We are specifying a 9 node cluster with 3 racks. cass-operator will evenly distribute C* 
+nodes across the racks.
+
+Each rack includes a zone property which corresponds to an availability zone. cass-operator 
+will use affinity and anti-affinity to ensure that pods from each rack are scheduled into 
+the corresponding availability zone and also to ensure that C* pods are not co-located on 
+k8s worker nodes.
+
+Cass Operator always configures the cluster to use `GossipingPropertyFileSnitch`. I am not 
+sure what happens if I try to specify a different value for `endpoint_snitch`, but 
+regardless, this is an example of a setting that we should not expose in k8ssandra.
+
+Now let's look at how to create a multi-DC cluster (multi-region is outside the scope of 
+this doc).
+
+Here is the first DC:
+
+```yaml
+apiVersion: cassandra.datastax.com/v1beta1
+kind: CassandraDatacenter
+metadata:
+  name: dc1
+  namespace: dev
+spec:
+  clusterName: multi-dc
+  ...
+```
+Here is the second DC:
+
+```yaml
+apiVersion: cassandra.datastax.com/v1beta1
+kind: CassandraDatacenter
+metadata:
+  name: dc2
+  namespace: dev
+spec:
+  clusterName: multi-dc
+  ...
+```
+
+We create two CassandraDatacenters that have the same clusterName and same namespace.
+
+If we want common settings across the two DCs, e.g., auth, heap settings, etc., we need
+to specify those settings for each CassandraDatacenter.
+
+## Configuration
+Now let's look at how we specify some settings from cassandra.yaml and jvm.options:
+
+```yaml
+apiVersion: cassandra.datastax.com/v1beta1
+kind: CassandraDatacenter
+metadata:
+  name: dc1
+# ...
+spec:
+# ...
+  # Everything under the config key is passed to the 
+  # cass-config-builder init container that runs on pod creation, and 
+  # then marshalled into config files.
+  config:
+    cassandra-yaml:
+      num_tokens: 8
+      memtable_allocation_type: heap_buffers
+      file_cache_size_in_mb: 1000
+      
+      authenticator: org.apache.cassandra.auth.PasswordAuthenticator
+      authorizer: org.apache.cassandra.auth.CassandraAuthorizer
+      role_manager: org.apache.cassandra.auth.CassandraRoleManager
+      
+      roles_validity_in_ms: 2000
+      permissions_validity_in_ms: 2000
+      credentials_validity_in_ms: 2000
+      
+      # How long before a node logs slow queries. Select queries that 
+      # take longer than this timeout to execute, will generate an
+      # aggregated log message, so that slow queries can be identified. 
+      # Set this value to zero to disable slow query logging. 
+      slow_query_log_timeout_in_ms: 500
+
+    jvm-options:
+      # Set the database to use 14 GB of Java heap
+      initial_heap_size: "14G"
```

Comment:
Note to this.. does the customer still set their own properties for the max-memory the container uses? The non-heap part. Should we add some sort of "best effort guess" to this configuration in the future?

Reply:
>does the customer still set their own properties for the max-memory the container uses?

Yes, this should still be configurable. CassandraDatacenter has a `resources` property for specifying CPU and memory resources for the pod. Here is an example:

```yaml
apiVersion: cassandra.datastax.com/v1beta1
kind: CassandraDatacenter
metadata:
  name: dc1
spec:
  clusterName: resources-example
  serverType: cassandra
  serverVersion: 3.11.7
  managementApiAuth:
    insecure: {}
  size: 3
  config:
    jvm-options:
      initial_heap_size: "800m"
      max_heap_size: "800m"
  resources:
    requests:
      cpu: 2
      memory: 2Gi
    limits:
      cpu: 2
      memory: 2G
  storageConfig:
    cassandraDataVolumeClaimSpec:
      storageClassName: standard
      accessModes:
      - ReadWriteOnce
      resources:
        requests:
          storage: 5Gi
```
We definitely want to expose this.

>Should we add some sort of "best effort guess" to this configuration in the future?

That is a great question, and something definitely worth considering. Suppose we add a property to specify the type of workload - ready heavy, write heavy, time series, etc. We could provide more tailored defaults based on what the user specifies. If the user does not specify a workload type, we fall back to the baseline or generic defaults.